File: 200520_C/vs/LinkedList.c


// 링크드 리스트 구현

#include <stdio.h>
#include <stdlib.h>
struct Node
{
    struct Node *next;
    int data;
};

int main()
{

    struct Node *head = malloc(sizeof(struct Node)); // 머리 노드 생성
    //머리 노드는 데이터를 저장하지 않음

    struct Node *node1 = malloc(sizeof(struct Node)); // 첫 노드 생성
    head->next = node1;                               // 머리 노드 다음은 첫번째 노드
    node1->data = 10;                                 //첫 노드에 10 저장

    struct Node *node2 = malloc(sizeof(struct Node)); //두번째 노드 생성
    node1->next = node2;
    node2->data = 20;

    node2->next = NULL;

    struct Node *curr = head->next; //연결 리스트 순회용 포인터에 첫 노드의 주소 저장
    while (curr != NULL)
    {
        printf("%d\n", curr->data);
        curr = curr->next;
    }

    free(node2);
    free(node1);
    free(head);

    return 0;
}

File: (20_0103)Algorithm/BinaryTree_Traversal/LinkedList.js
class Node {
  constructor(data, next = null) {
    this.data = data;
    this.next = next;
  }
}

const node1 = new Node(100);
console.log("node1: ", node1);

class LinkedList {
  constructor() {
    this.head = null;
    this.size = 0;
  }

  // Insert First Node
  insertFirst(data) {
    this.head = new Node(data, this.head);
    console.log("this.head: ", this.head);
    this.size++;
  }

  // Insert last node
  insertLast(data) {
    let node = new Node(data);
    let current;

    //If empty, make head
    if (!this.head) {
      this.head = node;
    } else {
      current = this.head;
      while (current.next) {
        current = current.next;
      }
      current.next = node;
    }
    this.size++;
  }

  // Insert at index

  insertAt(data, index) {
    // if index is out of range
    if (index > 0 && index > this.size) {
      return;
    }

    // if first index
    if (index === 0) {
      this.insertFirst(data);
      return;
    }

    const node = new Node(data);

    let current, previous;

    // set current to first
    current = this.head;
    console.log("current: ", current);
    let count = 0;
    while (count < index) {
      previous = current; //Node before index
      console.log("previous: ", previous);
      count++;
      current = current.next; //node after index
    }

    node.next = current;

    previous.next = node;

    this.size++;
  }

  // Get at index
  getAt(index) {
    let current = this.head;
    let count = 0;
    while (current) {
      if (count === index) {
        console.log(current.data);
      }
      count++;
      current = current.next;
    }

    return null;
  }

  // Remove at index
  removeAt(index) {
    if (index > 0 && index > this.size) {
      return;
    }

    let current = this.head;
    let previous;
    let count = 0;

    //Remove first
    if (index === 0) {
      this.head = current.next;
    } else {
      while (count < index) {
        count++;
        previous = current;
        current = current.next;
      }
      previous.next = current.next;
    }
    this.size--;
  }

  // Clear list
  clearList() {
    this.head = null;
    this.size = 0;
  }

  //  Print list data
  printListData() {
    let current = this.head;
    while (current) {
      console.log(current.data);
      current = current.next;
    }
  }
}

const ll = new LinkedList();

ll.insertFirst(100);
ll.insertFirst(200);
ll.insertFirst(300);
ll.insertLast(400);
ll.insertAt(500, 2);
ll.removeAt(0);
ll.clearList();
ll.printListData();
ll.getAt(2);
